### PR TITLE
Serializable Qobj + Pass cpp/projectq simulator tests

### PR DESCRIPTION
**Pull request by ajavadia**
I realized that the previous PR merge broke our c++ simulator and Project Q simulators. It was because our c++ simulator tests were flawed (they were actually using the Python backend!), and the ProjectQ simulator was not installed on Travis.

The problem was with the Qobj dictionary created. The value for the`backend` field was a `Backend` instance, but this is not valid JSON and those simulators that expect and decode JSON will throw errors. I made this point to the backend string, but included a copy of the instance in the QuantumJob wrapper. Going forward, this will be part of the Qobj object.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: test/python/test_qasm_sympy_simulator.py [hunk 1]
```diff
@@ -61,7 +61,7 @@ def setUp(self):
                          }
                      ]}
         self.q_job = QuantumJob(self.qobj,
-                                backend='local_sympy_qasm_simulator',
+                                backend=SympyQasmSimulator,
```
Comment (atilag) [inline test/python/test_qasm_sympy_simulator.py]: I guess you meant: `SympyQasmSimulator()` ?

File: qiskit/_quantumjob.py [hunk 2]
```diff
@@ -82,12 +80,10 @@ def __init__(self, circuits, backend=None,
             # circuits is actually a qobj...validate (not ideal but convenient)
             self.qobj = circuits
         else:
-            if not backend:
-                raise QISKitError('backend needs to be specified if '
-                                  'preformatted==True.')
             self.qobj = self._create_qobj(circuits, circuit_config, backend,
                                           seed, resources, shots, do_compile)
-        self.backend = self.qobj['config']['backend']
+        self.backend_name = self.qobj['config']['backend_name']
```
Comment (atilag) [inline qiskit/_quantumjob.py]: `self.backend_name` is  not being used elsewhere, right?

**Comment by atilag** [comment]
I'm dreaming with the day QObj is an actual class and not a dictionary :)

**Comment by jaygambetta** [comment]
@ajavadia will be soon.

File: test/python/test_qasm_sympy_simulator.py (hunk 1 above)
Comment (ajavadia) [reply test/python/test_qasm_sympy_simulator.py]: yep, thanks!

File: qiskit/_quantumjob.py (hunk 2 above)
Comment (ajavadia) [reply qiskit/_quantumjob.py]: yep. removed.

File: qiskit/_quantumjob.py [hunk 3]
```diff
@@ -82,12 +80,9 @@ def __init__(self, circuits, backend=None,
             # circuits is actually a qobj...validate (not ideal but convenient)
             self.qobj = circuits
         else:
-            if not backend:
-                raise QISKitError('backend needs to be specified if '
-                                  'preformatted==True.')
```
Comment (diego-plan9) [inline qiskit/_quantumjob.py]: Can you also remove the
```
         Raises:
             QISKitError: if preformatted==True but no backend is specified.
```

bit from the docstring, for consistency?

**Pull request merged**